osu.Game.Rulesets.Osu/Difficulty/Evaluators/ReadingEvaluator.cs

```diff
@@ -0,0 +1,241 @@
+        private const double hidden_multiplier = 0.28;
+        private const double density_multiplier = 2.4;
+        private const double density_difficulty_base = 2.5;
+        private const double preempt_balancing_factor = 140000;
+        private const double preempt_starting_point = 500; // AR 9.66 in milliseconds
+        private const double minimum_angle_relevancy_time = 2000; // 2 seconds
+        private const double maximum_angle_relevancy_time = 200;
+
+        public static double EvaluateDifficultyOf(int totalObjects, DifficultyHitObject current, double clockRate, double preempt, bool hidden)
+        {
+            if (current.BaseObject is Spinner || current.Index == 0)
+                return 0;
+
+            var currObj = (OsuDifficultyHitObject)current;
+            var nextObj = (OsuDifficultyHitObject)current.Next(0);
+            double constantAngleNerfFactor = getConstantAngleNerfFactor(currObj);
+            double velocity = Math.Max(1, currObj.LazyJumpDistance / currObj.AdjustedDeltaTime); // Only allow velocity to buff
+
+            double pastObjectDifficultyInfluence = getPastObjectDifficultyInfluence(currObj, preempt);
+
+            double noteDensityDifficulty = calculateDensityDifficulty(currObj, nextObj, velocity, constantAngleNerfFactor, pastObjectDifficultyInfluence, totalObjects, preempt);
+
+            double hiddenDifficulty = hidden ? calculateHiddenDifficulty(currObj, pastObjectDifficultyInfluence, totalObjects, preempt, velocity, clockRate, constantAngleNerfFactor) : 0;
+
+            double preemptDifficulty = calculatePreemptDifficulty(velocity, constantAngleNerfFactor, preempt);
+
+            double difficulty = DifficultyCalculationUtils.Norm(1.5, preemptDifficulty, hiddenDifficulty, noteDensityDifficulty);
+
+            return difficulty;
+        }
+
+        /// <summary>
+        /// Calculates the density difficulty of the current object and how hard it is to aim it because of it based on:
+        /// <list type="bullet">
+        /// <item><description>cursor velocity to the current object,</description></item>
+        /// <item><description>how many times the current object's angle was repeated,</description></item>
+        /// <item><description>the influence of the amount of previous objects visible on the screen when the current object appears</description></item>
+        /// </list>
+        /// </summary>
+        private static double calculateDensityDifficulty(OsuDifficultyHitObject currObj, OsuDifficultyHitObject? nextObj, double velocity, double constantAngleNerfFactor,
+                                                         double pastObjectDifficultyInfluence, int totalObjects,
+                                                         double preempt)
+        {
+            double futureObjectDifficultyInfluence = Math.Sqrt(retrieveCurrentVisibleObjects(totalObjects, currObj, preempt));
```

Comment:
Pass `retrieveCurrentVisibleObjects` result from the outside to not calculate twice